### ISSUE TITLE
Build fails on 32-bit linux 

### ISSUE THREAD

**Issue by MagaTailor** [ttsiodras/renderer#1]
The configure script correctly detects the processor has just SSE available but:

```
make[2]: Entering directory `/home/petevine/.phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.5.0/renderer-2.2z/src'
  CXX    renderer-renderer.o
In file included from Screen.h:41:0,
                 from renderer.cc:35:
Screen.h: In function 'DrawPixelBasic<4>(int, int, unsigned int)void':
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1449:1: error: inlining failed in call to always_inline '_mm_stream_si32(int*, int)': target specific option mismatch
 _mm_stream_si32 (int *__A, int __B)
 ^
In file included from renderer.cc:35:0:
Screen.h:357:6: error: called from here
      _mm_stream_si32((int*)bufp, (int)color);
      ^
make[2]: *** [renderer-renderer.o] Error 1
make[2]: Leaving directory `/home/petevine/.phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.5.0/renderer-2.2z/src'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/petevine/.phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.5.0/renderer-2.2z/src'
make: *** [all-recursive] Error 1
```

**Comment by ttsiodras** [comment]
Thanks for the bug report - I can see that this is happening in Slackware, but can you also provide the version of GCC that you are using? The output from `gcc --version` should suffice.

**Comment by MagaTailor** [comment]
Here you are:

```
Reading specs from /usr/lib/gcc/i586-slackware-linux/4.9.3/specs
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/i586-slackware-linux/4.9.3/lto-wrapper
Target: i586-slackware-linux
Configured with: ../gcc-4.9.3/configure --prefix=/usr --libdir=/usr/lib --mandir=/usr/man --infodir=/usr/info --enable-shared --enable-bootstrap --enable-languages=ada,c,c++,fortran,go,java,lto,objc --enable-threads=posix --enable-checking=release --enable-objc-gc --with-system-zlib --with-python-dir=/lib/python2.7/site-packages --disable-libunwind-exceptions --enable-__cxa_atexit --enable-libssp --enable-lto --enable-install-libiberty --with-gnu-ld --verbose --enable-java-home --with-java-home=/usr/lib/jvm/jre --with-jvm-root-dir=/usr/lib/jvm --with-jvm-jar-dir=/usr/lib/jvm/jvm-exports --with-arch-directory=i386 --with-antlr-jar=/root/slackware-current/source/d/gcc/antlr-runtime-3.4.jar --enable-java-awt=gtk --disable-gtktest --with-arch=i586 --target=i586-slackware-linux --build=i586-slackware-linux --host=i586-slackware-linux
Thread model: posix
gcc version 4.9.3 (GCC)
```

I believe I saw a similar error trying to build on arm a while ago but I must have dismissed the whole idea after seeing intrinsics in there :)

**Comment by ttsiodras** [comment]
Can you also paste the output of `cat /proc/cpuinfo  | grep ^flags | head -1`?

I need to see what level of SSE support your machine has.

**Comment by MagaTailor** [comment]
Just SSE, which I mentioned in the opening sentence :)

```
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 mmx fxsr sse syscall mp mmxext 3dnowext 3dnow vmmcall

```

**Comment by ttsiodras** [comment]
Thanks, @petevine .  Can you try it now, with the new version? I uploaded it on my site as well, so phoronix tests should be able to use it - unless Michael does an MD5 checksum check on the tarball contents... in which case I guess an update is needed on Michael's repo.

**Comment by MagaTailor** [comment]
Got a little further but still no luck:

```
CXX      renderer-MLAA.o
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
MLAA.cc: In function 'ssedif(long long __vector&, long long __vector&)':
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:645:1: error: inlining failed in call to always_inline '_mm_set1_epi8(char)': target specific option mismatch
 _mm_set1_epi8 (char __A)
 ^
MLAA.cc:50:61: error: called from here
   __m128i hibits = _mm_set1_epi8(most_significant_color_bits);
                                                             ^
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1380:1: error: inlining failed in call to always_inline '_mm_min_epu8(long long __vector, long long __vector)': target specific option mismatch
 _mm_min_epu8 (__m128i __A, __m128i __B)
 ^
MLAA.cc:51:70: error: called from here
   __m128i d = _mm_sub_epi8(_mm_max_epu8(c0, c1), _mm_min_epu8(c0, c1));
                                                                      ^
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1368:1: error: inlining failed in call to always_inline '_mm_max_epu8(long long __vector, long long __vector)': target specific option mismatch
 _mm_max_epu8 (__m128i __A, __m128i __B)
 ^
MLAA.cc:51:70: error: called from here
   __m128i d = _mm_sub_epi8(_mm_max_epu8(c0, c1), _mm_min_epu8(c0, c1));
                                                                      ^
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1055:1: error: inlining failed in call to always_inline '_mm_sub_epi8(long long __vector, long long __vector)': target specific option mismatch
 _mm_sub_epi8 (__m128i __A, __m128i __B)
 ^
MLAA.cc:51:70: error: called from here
   __m128i d = _mm_sub_epi8(_mm_max_epu8(c0, c1), _mm_min_epu8(c0, c1));
                                                                      ^
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1264:1: error: inlining failed in call to always_inline '_mm_and_si128(long long __vector, long long __vector)': target specific option mismatch
 _mm_and_si128 (__m128i __A, __m128i __B)
 ^
MLAA.cc:52:31: error: called from here
   d = _mm_and_si128(d, hibits);
                               ^
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:749:1: error: inlining failed in call to always_inline '_mm_setzero_si128()': target specific option mismatch
 _mm_setzero_si128 (void)
 ^
MLAA.cc:53:44: error: called from here
   d = _mm_cmpeq_epi8(d, _mm_setzero_si128());
                                            ^
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1288:1: error: inlining failed in call to always_inline '_mm_cmpeq_epi8(long long __vector, long long __vector)': target specific option mismatch
 _mm_cmpeq_epi8 (__m128i __A, __m128i __B)
 ^
MLAA.cc:53:45: error: called from here
   d = _mm_cmpeq_epi8(d, _mm_setzero_si128());
In file included from /usr/lib/gcc/i586-slackware-linux/4.9.3/include/xmmintrin.h:1258:0,
                 from MLAA.cc:26:
/usr/lib/gcc/i586-slackware-linux/4.9.3/include/emmintrin.h:1386:1: error: inlining failed in call to always_inline '_mm_movemask_epi8(long long __vector)': target specific option mismatch
 _mm_movemask_epi8 (__m128i __A)
 ^
MLAA.cc:54:45: error: called from here
   unsigned short int f = _mm_movemask_epi8(d);
                                             ^
make[2]: *** [renderer-MLAA.o] Error 1
make[2]: Leaving directory `/home/petevine/.phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.5.0/renderer-master/src'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/petevine/.phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.5.0/renderer-master/src'
make: *** [all-recursive] Error 1
```

**Comment by ttsiodras** [comment]
Are you forcing MLAA? If you are passing `--enable-mlaa` to `configure`, then you are forcing use of Intel's MLAA, which depends on SSE2:

```
configure: ************************************************************
configure: ** Intel MLAA antialiasing is disabled! use --enable-mlaa **
configure: ** (provided your CPU has SSE2 instructions...            **
configure: ************************************************************
...
```

**Comment by MagaTailor** [comment]
Definitely not, the default is off and the message is screaming about SSE2 so it's the last thing I'd want to force.

**Comment by ttsiodras** [comment]
Are you sure? Maybe it's a remnant of a previous build, but from `src/Makefile.am` you can clearly see...

```
renderer_SOURCES = renderer.cc ${common_SRC}
#if MLAA_ENABLED
renderer_SOURCES += MLAA.h MLAA.cc
#endif
```

...that the only way MLAA.cc is added to the set of files to build, is if MLAA_ENABLED is defined.
And from `configure.ac`:

```
...
if test x"${enable_mlaa}" = xyes ; then
    AC_DEFINE([MLAA_ENABLED], 1, [Define this to use Intel SSE2 implementation of MLAA.])
else
    AC_MSG_NOTICE([************************************************************])
    AC_MSG_NOTICE([** Intel MLAA antialiasing is disabled! use --enable-mlaa **])
    AC_MSG_NOTICE([** (provided your CPU has SSE2 instructions...            **])
    AC_MSG_NOTICE([************************************************************])
fi
...
```

**Comment by ttsiodras** [comment]
To verify that MLAA is indeed not defined - do you see this kind of output?

```
$ grep MLAA src/config.h
/* Define this to use Intel SSE implementation of MLAA. */
/* #undef MLAA_ENABLED */
```

**Comment by MagaTailor** [comment]
There was no previous build as I was compiling in a clean master source directory so it must be a configure issue.

As for MLAA it's exactly what I've got in my config.h - maybe it should get undefined explicitly?

**Comment by ttsiodras** [comment]
Hmm... No, the errors you pasted clearly tell us that the Makefile generated by configure on your machine includes `MLAA.cc` - even though `configure` has clearly (based on the output of the `grep`) told it NOT to.

Very weird...  Note that I've tested the build process not only under Linuxes, but also under the BSDs  (and even OS/X). Something is amiss under Slackware, but I'd have to install it to see...

Anyway, as a workaround - so you can build the renderer and see it run - just remove MLAA.cc from src/Makefile.am:

```
diff --git a/src/Makefile.am b/src/Makefile.am
index 0ac5cdd..40c865b 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -9,7 +9,7 @@ common_SRC = \

 renderer_SOURCES = renderer.cc ${common_SRC}
 #if MLAA_ENABLED
-renderer_SOURCES += MLAA.h MLAA.cc
+#renderer_SOURCES += MLAA.h MLAA.cc
 #endif
```

...and do the usual `autoreconf ; automake ; ./configure ; make`

If after this, the Makefile generated by configure still includes `MLAA.cc` ... then your machine is haunted :-)

**Comment by MagaTailor** [comment]
Yep, it compiles now. Thx. 

Did you ever test it on non-SSE2 machines? Cause it's probably an issue with the build scripts/make tools somewhere and may have stayed undetected due to the availability of SSE2 even with MLAA disabled.

**Comment by ttsiodras** [comment]
The issue you have is that the generated Makefile under Slackware includes `MLAA.cc` even though configure explicitly says not to.  I just tested this aspect of the build in my machine, and also in a Docker-ized Ubuntu 12. In both cases, the `src/Makefile`s generated with `--enable-mlaa` and without it, behave as expected - i.e. they either include `MLAA.cc` (and therefore compile it) or they don't.

Somehow under Slackware, the  `#if MLAA_ENABLED` is ignored...  No idea why, maybe we should ask Patrick for help :-)

**Comment by MagaTailor** [comment]
It's not about Slackware per se but probably about certain versions of automake tools that are present at the moment. (upgraded over time)

**Comment by MagaTailor** [comment]
It seems I was absolutely right about the tools - I'm getting the same behaviour (MLAA.o being built) on an armv7 Ubuntu 14.04 system.

This probably means your auto tools scripts are slightly outdated :)

**Comment by ttsiodras** [comment]
Maybe - but I don't think so. I am more inclined to believe that the pattern I used (the `#if` inside `Makefile.am`) is not portable, and doesn't work on all UNICes. 

I'll look into it - thanks for bringing it to my attention! In fact since this is a different bug, you may choose to open a new ticket for this.
